DRILL-8028: Add PDF Format Plugin

cgivre opened the pull request:
# [DRILL-8028](https://issues.apache.org/jira/browse/DRILL-8028): Add PDF Format Plugin

## Description
This PR adds the capability to query data contained in PDF tables. 

## Documentation
One of the most annoying tasks is when you are working on a data science project and you get data that is in a PDF file. This plugin endeavours to enable you to query data in
 PDF tables using Drill's SQL interface.  

## Data Model
Since PDF files generally are not intended to be queried or read by machines, mapping the data to tables and rows is not a perfect process.  The PDF reader does support 
provided schema. 

### Merging Pages
The PDF reader reads tables from PDF files on each page.  If your PDF file has tables that span multiple pages, you can set the `combinePages` parameter to `true` and Drill 
will merge all the tables in the PDF file.  You can also do this at query time with the `table()` function.

## Configuration
To configure the PDF reader, simply add the information below to the `formats` section of a file base storage plugin. 

```json
"pdf": {
  "type": "pdf",
  "extensions": [
    "pdf"
  ],
  "extractHeaders": true,
  "combinePages": false
}
```
The avaialable options are:
* `extractHeaders`: Extracts the first row of any tables as the header row.  If set to false, Drill will assign column names of `field_0`, `field_1` to each column.
* `combinePages`: Merges multipage tables together.
* `defaultTableIndex`:  Allows you to query different tables within the PDF file. Index starts at `0`. 


## Accessing Document Metadata Fields
PDF files have a considerable amount of metadata which can be useful for analysis.  Drill will extract the following fields from every PDF file.  Note that these fields are not
 projected in star queries and must be selected explicitly.  The document's creator populates these fields and some or all may be empty. With the exception of `_page_count
 ` which is an `INT` and the two date fields, all the other fields are `VARCHAR` fields.
 
 The fields are:
 * `_page_count`
 * `_author`
 * `_title`
 * `_keywords`
 * `_creator`
 * `_producer`
 * `_creation_date`
 * `_modification_date`
 * `_trapped`
 * `_table_count`
 
 The query below will access a document's metadata:
 
 ```sql
SELECT _page_count, _title, _author, _subject, 
_keywords, _creator, _producer, _creation_date, 
_modification_date, _trapped 
FROM dfs.`pdf/20.pdf`
```
The query below demonstrates how to define a schema at query time:

```sql
SELECT * FROM table(cp.`pdf/schools.pdf` (type => 'pdf', combinePages => true, 
schema => 'inline=(`Last Name` VARCHAR, `First Name Address` VARCHAR, 
`field_0` VARCHAR, `City` VARCHAR, `State` VARCHAR, `Zip` VARCHAR, 
`field_1` VARCHAR, `Occupation Employer` VARCHAR, 
`Date` VARCHAR, `field_2` DATE properties {`drill.format` = `M/d/yyyy`}, 
`Amount` DOUBLE)')) 
LIMIT 5
```

## Testing
Added unit tests

luocooong commented:
@cgivre Thanks for the new format plugin. Is it ready for review?

cgivre commented:
> @cgivre Thanks for the new format plugin. Is it ready for review?

This should be ready for review :-)

paul-rogers commented:
Cool contribution. I'm not entirely convinced this is something that Drill should handle. There are too many variables for the very limited controls which Drill provides. It is likely that this will work for one or two limited use cases, but not the vast majority of PDF files. Using JSON plugin config files to specify the mapping is awkward. Probably each file will need its own config, which is not scalable.

Drill's fundamental design is to run at scale. It is highly unlikely that someone will use PDF files to store GB of data. If they do, they have problems bigger than Drill can help them solve. Thus, this kind of plugin works only at the small scale: one or two files in, say, an embedded Drillbit with JDBC or SQLine.

A better choice would be to wrap this thing in a script: tinker with the PDF extraction, using whatever tools are available, to get the right mapping. Then, wrap that in a script that produces, say, a CSV format to stdout. Drill can then read that input.

Such an approach enables all manner of ad-hoc, small scale data extraction.

Or, maybe Drill should offer a "desktop edition" that is designed for small, ad-hoc projects based on local files, with some way to handle all the tinkering needed when reading PDF files, images, Word files, spreadsheets, Twitter feeds, Slack posts another formats popular with data scientists. Such features would not normally be part of the massive-scale deployments for which Drill is designed.

Thoughts?

cgivre commented:
@paul-rogers  Thanks for the feedback.  I was on the fence about submitting this as well.  I was kind of doing a "branch purge" and wanted to contribute some experiments I had been poking at.  Let me first give you the backstory...  Several years ago, I worked on an analytics project which unfortunately, we had to scrap because the data we received was in PDF format.   While there are python libraries that can parse PDFs, (and probably others), many have C dependencies which we could not build in our environment.  This project weighed on me because had we been able to actually read the data, we could have solved a mission critical issue.   

Fast forward a few years, I got into Drill, learned how to write format plugins, and stumbled on the Tabula library which is built for the purpose of extracting and parsing tables from PDF files.   I wanted to see if Drill could read these and if so, how well it did at extracting the tables and getting usable data from them.   The actual result was that it worked surprisingly well.   The extractors here do a very good job of identifying tables and extracting them into a tabular structure.

OK... see inline for the rest.

> Cool contribution. I'm not entirely convinced this is something that Drill should handle. 

I wasn't so sure either.  However, we do already support file types which are not used for big data processing such as ESRI shape files, and a few other random formats.  If you are viewing Drill solely as a query engine for big data, then yes this makes no sense.  If you are viewing Drill as a Swiss Army knife for data, then it makes more sense.

> There are too many variables for the very limited controls which Drill provides. It is likely that this will work for one or two limited use cases, but not the vast majority of PDF files. 

Let me clarify that this is only for PDF files which contain tables.  (Either images that look like tables or actual text tables).  By design this will not work with all PDF files.  

> Using JSON plugin config files to specify the mapping is awkward. Probably each file will need its own config, which is not scalable.  

I'm not sure what you're getting at here with the JSON config.  Could you be mixing this up with another recent PR which did add some JSON config options to the HTTP/REST storage plugin?  In any event, this format reader only has two config variables.  One which merges all the tables in the document together (in the case of a table which spans multiple pages), and the other which specifies the index of the table to parse in the event of a document with multiple tables that are not related.

> Drill's fundamental design is to run at scale. It is highly unlikely that someone will use PDF files to store GB of data. If they do, they have problems bigger than Drill can help them solve. 
Indeed!  

> Thus, this kind of plugin works only at the small scale: one or two files in, say, an embedded Drillbit with JDBC or SQLine.
> 
> A better choice would be to wrap this thing in a script: tinker with the PDF extraction, using whatever tools are available, to get the right mapping. Then, wrap that in a script that produces, say, a CSV format to stdout. Drill can then read that input.

True...  The issues that I ran into were that it was surprisingly difficult to get the libraries that can parse PDFs to run.  

> 
> Such an approach enables all manner of ad-hoc, small scale data extraction.
> 
> Or, maybe Drill should offer a "desktop edition" that is designed for small, ad-hoc projects based on local files, with some way to handle all the tinkering needed when reading PDF files, images, Word files, spreadsheets, Twitter feeds, Slack posts another formats popular with data scientists. Such features would not normally be part of the massive-scale deployments for which Drill is designed.
> 
> Thoughts?

I actually really do like the idea of a mini-drill.  For smaller use cases but without all the overhead and complexity.   

With that said, I don't really see a harm of including this.  Clearly this is not Drill's primary use case, but it can be really valuable.  I was actually helping my wife the other day with some data for her work.   In that situation, a state government agency sent her data in PDF format which we had to enter into a website.  Using Drill we were able to query the file and transform it into something actually useful.

dzamo commented:
@cgivre @paul-rogers, my 2c.  I guess some partial precedents for a format plugin like this are ones like format-image and format-esri (as noted), though those do only go after the explicitly structured content.  It would not surprise me if there are quite often cases of unfortunate people needing to scrape data out of 10^0 - 10^5 PDFs.  The purist in me agrees with Paul's thought: a Groovy script over whatever Java lib is used here could be employed instead.  That would not be automatically be parallelised like a Drill query is so a user with many PDFs _might_ be pushed all the way through GNU parallel to Spark.  All that is followed by this recurring thought: if Drill is disciplined and focussed only on SQL against standard big data formats then it finds itself trying to compete in an uncomfortably crowded space.  Probably fatally crowded.

I do also note that the "big" and "small" data worlds are not disjoint.  I have in practice joined big data in Parquet with small reference data in Excel (actually EtherCalc).  Even in the big data regime reference data remains small and is maintained by humans in human forrmats rather than pumped out by machines in machine formats.

This is getting long again.  Last thoughts.

- I feel this really should be good at finding and parsing tables (work often, rather than work seldom) for inclusion.
- We should consider a subproject to contain our long tail of "non-standard" data formats.  This would separate away plugins that should not be expected to run with speed or realiability of the core data formats and keep the core distributable size down as we add format after format.  We could then start to distribute tarballs for `drill-core` and `drill-extra`.
- This looks like a plugin that will benefit from Drill's optional schema.  That means we might have some unique ability to compete here.

paul-rogers commented:
@cgivre, @dzamo raise good points. So, what is Drill today? Is it still primarily used for distributed queries at scale? Or, as a handy desktop tool for data scientists? Probably both. The problem is, a tool that tries to be both a desert topping and a floor wax (let's see how old the readers are with this one), ends up being good at neither.

One approach, if we had resources, would be to create a Drill Desktop that is optimized for that case and encourages all kinds of specialized data connectors. Create an easy way to define those connectors (YAML files created by specialized web apps?) Ensure Drill has good integration with Jupyter and the other usual suspects.

Another approach, if we had resources, is the oft-discussed idea of separating the less-common plugins from the Drill core. Work started on this: to create an extension mechanism that made this possible. (Today, most plugins need quite a bit of Drill internal code.)

So, no harm in adding the PDF reader, but I expect usage will be pretty limited just because, for the folks that need it, configuration will be too hard. Better would be a Python or Spark job that extracts the data into a CSV file, then query the CSV file with Drill. Each step could be debugged easily. I can't imagine anyone will want to debug their PDF extraction using Drill's overly generous Java stack traces...

dzamo commented:
>  The problem is, a tool that tries to be both a desert topping and a floor wax (let's see how old the readers are with this one), ends up being good at neither.

@paul-rogers you got me with this idiom, but I like it!  The broader topic is super interesting.  If SQLite started adding the features needed to compete with Oracle Database 21c it would quickly fail at being SQLite.  If Linux tried to be an OS kernel for both TVs and supercomputers it would... continue to dominate both extremes!  There are some twists here!

Pigeonholing formats into small scale and large scales is also a tricky business.  For example, we naturally want to declare PDF a desktop format, but I can easily imagine a conversation like the following.

"Hey Bob, remember that we sent decades of paper archives from the basement out to that big scanning centre for digitisation?  They've come back as millions of pages of PDFs.  Someone just asked me if we can help them find all invoices containing a particular SKU, and pull out the price on that line.  The ERP system only has the last 10 years loaded into it and they want to go back further".

"Chuck 'em in HDFS, we'll run a Drill query"

"But PDF is a desktop publishing format, not a big data format!  Surely our big data cluster will want nothing to do with it!?"

"Drill's got a plugin architecture which led to people adding support for all sorts of weird and wonderful formats.  Querying PDFs is a dubious business but we'll know after ~10 lines of SQL if we can do this with Drill or not.  If not, miserable days or weeks of programming with a PDF library await one of our interns."

paul-rogers commented:
@dzamo, the reference was to an early Saturday Night Live skit.

Just to refocus the discussion, my question is really around configuration. When querying CSV, Parquet and the like, it is very clear where the data lies. When querying Excel, PDF, HTML, Word, "the web", it is less clear: there is some amount of data mining needed to say that it is THIS table and not THAT one, that the table spans three pages, etc.

The question is, how does the user specify this? If it were me, I would not want to be tinkering with a JSON storage plugin config, watching my query fail with a Drill stack trace, or wondering why I got no data. Instead, I'd want a tool better suited for the task. Once I had that, if I then wanted to run at scale, I'd want to say, "Drill, just use X and consume the data."

So, the question here is: is the JSON storage plugin config an effective way for a data scientist to mine PDF, Excel, HTML and other messy sources? I don't have an answer, I'm just asking the question.

Again, if we had the ability to have true external plugins, then this could easily be an add-on project. Those who know PDF could go hog wild creating a good solution. But, we don't, so every specialized plugin has to be part of core Drill. Is that good? That's also a question for debate.

dzamo commented:
@paul-rogers, right, okay it's an expressiveness thing here rather than a scale thing.  The expressiveness of Drill SQL ∪ Drill format config JSON falls well short of that of a general purpose scripting language and for reading fiddly unstructured data that shortfall might rapidly become uncomfortable.  The format config for this particular plugin looks quite succinct, like the plugin will either automagically get your data out, or it won't and then you need to pack up and go and open the interpreter of your favourite scripting language.  Making your resulting script scale to millions of pages, if it that's needed, is left to the student.  I quite like the Ray project for Python myself.

This thread has triggered some thoughts.  If we find ourselves starting to write long essays of JSON in format configs then we should probably be concerned.  If we find ourselves trying to embed a miniature data processing DSL into format config JSON then we need to stop moving immediately and pray to the ancestors that we might be shown a path that will return us from wilderness.  I want to revisit the draft fixed width format plugin with these ideas in mind.  Its config allows setting names and types for columns, but for other formats we must do this in SQL.  I think we should only ever do this in SQL.

I think we can do something on the packaging front.  These format plugins live under contrib/ in the source tree and are compiled to their own jar files.  If we simply change the final tarball-building stage of our Maven build to give us something like the following on our download page, would we not be in reasonable shape? 

Package|Size|Description
--|--|--
drill-core|300MB|Drill with core storage layer libs only.  Use this in a focussed big data environment to query standard formats like Parquet, CSV and JSON in HDFS or object storage with predictable results and performance.  Supplement this with indiviudal plugins listed below as needed.
drill-ktichen-sink|1.5GB|Drill core plus all 100+ storage and format plugins.  Use this for maximum compatibility.  Results and performance may vary across plugins.
drill-storage-jdbc|130KB|Plugin to query systems that provide a JDBC driver using a generic SQL dialect.
drill-format-pdf|90KB|Plugin to query tables scraped from PDF files.
...

P.S. We'd be persisting with a monolithic Git repo containing multiple "projects" here, but I personally don't mind mono repos.

dzamo commented:
@cgivre Instead of generating `field_0`, `field_1` when no column names can be determined, is it possible to generate a `columns` array to be accessed using `columns[0], columns[1], ...` to remain consistent with what we do for CSV?

cgivre commented:
> @cgivre Instead of generating `field_0`, `field_1` when no column names can be determined, is it possible to generate a `columns` array to be accessed using `columns[0], columns[1], ...` to remain consistent with what we do for CSV?

It's possible to do that, but a few things:
1.  Sometimes the PDF reader does not read tables perfectly and you get a mix of found headers and not found headers, so that's one reason I took that approach.
2.  I actually dislike the `columns` approach from the CSV readers because it increases the level of complexity of queries.  In theory, if someone is querying a table (doesn't matter from where) they will want that broken into columns and rows.  The columns array approach (IMHO) makes this a lot harder that it needs to be.  
3. This actually follows the model used in the Excel reader.

cgivre commented:
> @dzamo, the reference was to an early Saturday Night Live skit.
> 
> Just to refocus the discussion, my question is really around configuration. When querying CSV, Parquet and the like, it is very clear where the data lies. When querying Excel, PDF, HTML, Word, "the web", it is less clear: there is some amount of data mining needed to say that it is THIS table and not THAT one, that the table spans three pages, etc.
> 
> The question is, how does the user specify this? If it were me, I would not want to be tinkering with a JSON storage plugin config, watching my query fail with a Drill stack trace, or wondering why I got no data. Instead, I'd want a tool better suited for the task. Once I had that, if I then wanted to run at scale, I'd want to say, "Drill, just use X and consume the data."
> 
> So, the question here is: is the JSON storage plugin config an effective way for a data scientist to mine PDF, Excel, HTML and other messy sources? I don't have an answer, I'm just asking the question.
> 
> Again, if we had the ability to have true external plugins, then this could easily be an add-on project. Those who know PDF could go hog wild creating a good solution. But, we don't, so every specialized plugin has to be part of core Drill. Is that good? That's also a question for debate.

@dzamo @paul-rogers , These are all interesting points.  I'd like to focus on the config variables for a moment.  The PDF reader has 2 config variables:  the `mergeAllTables` and the `tableNumber`.   While a user could theoretically set these globally, they really are intended to be set via the `table()` function at query time.   For some reference the Excel reader has similar configs which allow a user to select different sheets within an Excel file, or define regions in a file where their data lives etc. 

IMHO, this flexibility is actually very good for the user, because it allows an administrator to configure global default values that make sense but also allow a user to make query-time changes so they can access their data.

With respect to the fixed-width plugin (https://github.com/apache/drill/pull/2282) I actually have a different vision of how this can be used, and will post comments there.

dzamo commented:
> * Sometimes the PDF reader does not read tables perfectly and you get a mix of found headers and not found headers, so that's one reason I took that approach.

This consideration will only apply for `extractHeaders = true`, right?  In this case all of the readers do split out columns so I think we're good.
 
> * I actually dislike the `columns` approach from the CSV readers because it increases the level of complexity of queries.  In theory, if someone is querying a table (doesn't matter from where) they will want that broken into columns and rows.  The columns array approach (IMHO) makes this a lot harder that it needs to be.

The columns array does allow text files to contain jagged arrays, which is perhaps valuable?  I don't see a huge saving in typing `select field_0, field_1` over `select columns[0], columns[1]`.

> * This actually follows the model used in the Excel reader.

My fear, and I don't know if this is real or not, is that if we profilerate plugin-specific quirks in how the schema is represented to the user then the promise of a standard SQL interface to the data gets tainted and developers and users will be put off.  

"It's standard except every plugin presents its data the way that author prefers"
"That fragment of code I sent you for the columns array won't work here because this plugin does it differently"
"This SQL script makes for confusing reading because the plugins involved name their generated columns differently, sometimes 'column', sometimes 'field', sometimes 'var'"
etc.

I don't feel all that strongly about `field_0` vs `columns[0]` (I mean, maybe we deprecate the columns array?) but I am finding myself thinking more and more about consistency.

cgivre commented:
> > * Sometimes the PDF reader does not read tables perfectly and you get a mix of found headers and not found headers, so that's one reason I took that approach.
> 
> This consideration will only apply for `extractHeaders = true`, right? In this case all of the readers do split out columns so I think we're good.
> 
> > * I actually dislike the `columns` approach from the CSV readers because it increases the level of complexity of queries.  In theory, if someone is querying a table (doesn't matter from where) they will want that broken into columns and rows.  The columns array approach (IMHO) makes this a lot harder that it needs to be.
> 
> The columns array does allow text files to contain jagged arrays, which is perhaps valuable? I don't see a huge saving in typing `select field_0, field_1` over `select columns[0], columns[1]`, am I missing some other complication?
> 
> > * This actually follows the model used in the Excel reader.
> 
> My fear, and I don't know if this is real or not, is that if we profilerate plugin-specific quirks in how the schema is represented to the user then the promise of a standard SQL interface to the data gets tainted and developers and users will be put off.
> 
> "It's standard except every plugin presents its data the way that author prefers" "That fragment of code I sent you for the columns array won't work here because this plugin does it differently" "This SQL script makes for confusing reading because the plugins involved name their generated columns differently, sometimes 'column', sometimes 'field', sometimes 'var'" etc.
> 
> I don't feel all that strongly about `field_0` vs `columns[0]` (I mean, maybe we deprecate the columns array?) but I am finding myself thinking more and more about consistency.

The bigger implication of having a `columns` array vs `field_n` is when a user starts with `SELECT *` queries.  It makes it harder for BI tools to gather schema metadata and it also is non-standard SQL.  Now... querying PDF is also non-standard SQL... so maybe that's less important.  But, it makes the discovery a little harder IMHO.

dzamo commented:
> The bigger implication of having a `columns` array vs `field_n` is when a user starts with `SELECT *` queries. It makes it harder for BI tools to gather schema metadata and it also is non-standard SQL. Now... querying PDF is also non-standard SQL... so maybe that's less important. But, it makes the discovery a little harder IMHO.

@cgivre Oh right, that makes sense.  So should we put in support for both `columns[n]` and `field_n` as widely as possible, with a standardised option which lets users switch between each mode?  Maybe standardising on naming of `columns[n]` and `column_n` is a small saving on cognitive load for users here?

@paul-rogers with apologies to this PR for saddling it with so much broader design chat, I wanted to share a last set of findings from talking with others with you and finally ask if we might go over a couple of questions with you, away from this PR.

1. I've polled some Drill devs and going after the "long tail" of formats and storage systems is mostly of interest to them.  @vvysotskyi even has an intriguing idea of a marketplace for these plugins, I guess something like the Eclipse plugin marketplace.
2. I have developed a conviction that to go after the "long tail" and not produce a sprawling mess that neither developers nor users want to touch, we need to try to get strict (to the extent possible) about consistency in how plugins behave and how they are configured.  Today we already are not all that consistent (e.g. see remarks on `columns[n]` vs `field_n` above, on column `name` and `type` in fixed width format).
3. Those I've spoken with do also like the idea of splitting our distributed packages into "core" and "kitchen sink", or something like that, to put us in a better position to go after the "long tail".  It sounds like we're okay with our existing mono repo containing many plugins but end users should not have to download the kitchen sink to query e.g. just JSON or Parquet.  Drill startup times will probably be slow for the kitchen sink because the Java class loader will have a huge amount to scan.  And developer testing could get onerous if we cannot compile only a subset.
4. By chance I saw that BigQuery, which for some reason I've designated in my mind of as kind of the Rolls Royce of Dremel-family engines even though I know little about it, can query Google Sheets.  So even they entertain some "small data" formats, although nothing like what we're imagining.  Just an anecdote.

I would love to consult with you on 2 and 3 in a sort of "Very well, if you _must_ do a distribution of Drill with this long tail of formats, storage systems and UDFs in it then at least equip yourselves with the following practices" chat.  Perhaps in the upcoming community meetup, otherwise outside (if it's of any interest on your end of course).

Thanks
James

pjfanning commented on distribution/src/assemble/component.xml:
```diff
@@ -49,7 +49,8 @@
         <include>org.apache.drill.contrib:drill-format-hdf5:jar</include>
         <include>org.apache.drill.contrib:drill-format-ltsv:jar</include>
-        <include>org.apache.drill.contrib:drill-format-httpd:jar</include>
+        <include>org.apache.drill.contrib:drill-format-https:jar</include>
```
the httpd to https change looks like an unintentional change

cgivre replied:
Good catch!

luocooong reviewed (commented):
@cgivre It is a good solution for **drill** the table data in PDF. Please take note of my comments.

paul-rogers commented:
@dzamo, @cgivre: we should certainly encourage people to add plugins (connectors) to everything under the sun. The question really is: does the rather limited Drill core team want to support all of them at the same level as the main ones that Just Gotta Work. For huge files. On local, S3, HDFS. With splits. With metadata. With excellent error reporting. And on and on.

The PDF thing seems like the prototypical "not in the core, but a great add-on" example. Yes, some one user might want to scan zillions of PDF files on an HDFS system. But, the core Drill team can't spend its resources on that when there are still basics to address.

Let's do this: let's get this into Drill for now. Then:

* Ensure we have a good way to build plugins separate from the Drill code. (Last time I tried, I couldn't get it to work.)
* Explain how to create a plugin in a users own repo, built against Drill.
* Explain how to drop the plugin into a running Drill to add that functionality.

This way, even without a marketplace, folks can build their own connectors separate from the Drill repo.

A next step might be to split Drill's monorepo into multiple parts, as do many other projects. The current repo would be the core. Then, the contrib stuff could move to its own repo(s) under the Drill project. And, users can create their own extensions in their own repos.

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+    return builder.buildSchema();
+  }
+
+  private void buildWriterList() {
```
We build the writers two ways depending on provided schema or not. With V2, the idea is to build the schema once, give it to the `SchemaNegotiator`, and get back a writer with all the needed columns. Adding columns to a writer is a bit old-school, needed only for formats, such as JSON, where we really don't know the schema until we read a record.

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+
+      switch (type) {
+        case VARCHAR:
+          writers.add(new StringPdfColumnWriter(counter, fieldName, rowWriter));
```
V2 does this for you: builds writers based on the declared field type.

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+
+    @Override
+    public void load(RectangularTextContainer<?> cell) {
+      writer.setLong(Long.parseLong(cell.getText()));
```
First, bravo for using this trick to optimize the innermost per-column load path.

What happens if the parse fails? Need to catch the number format exception and convert it to a `UserException`. Or, you could let the EVF do it for you with its column converters which do exactly this work.

Aside: if this were a quick & dirty add-on, I'd not worry about error handling or performance. But, because it is part of core Drill, and Drill must scale, we have to worry about all these details, even if maybe no one will ever read a zillion huge PDFs. Because, if I don't, how do I then turn around and ask the next reader to provide a scalable solution?

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+    public void load(RectangularTextContainer<?> cell) {
+      LocalDate localDate;
+      if (Strings.isNullOrEmpty(this.dateFormat)) {
+       localDate = LocalDate.parse(cell.getText());
```
Nit: missing leading space.

cgivre replied:
Fixed

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfFormatConfig.java:
```diff
@@ -0,0 +1,117 @@
+                         @JsonProperty("combinePages") boolean combinePages,
+                         @JsonProperty("defaultTableIndex") int defaultTableIndex) {
+    this.extensions = extensions == null
+      ? Collections.singletonList("pdf")
```
`extractionAlgorithm` is not described in the README.

cgivre replied:
I removed that.  I thought it would be a good config option.  It didn't seem to make a difference.

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfFormatConfig.java:
```diff
@@ -0,0 +1,117 @@
+    return extractHeaders;
+  }
+
+  @JsonProperty("extractionAlgorithm")
```
`@JsonInclude(Include.NON_NULL)`?

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfFormatConfig.java:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  @JsonProperty("defaultTableIndex")
+  public int getDefaultTableIndex() { return defaultTableIndex; }
```
Not sure I like this name. What is a `defaultTableIndex`? I'm specifying it, right? So, it's not actually a "default." It is a `tableIndex`. The default is what we get if I don't specify anything.

If I don't specify anything I get... all tables? So, I can either get one table or all tables? What if I want tables 1, 5 and 20-32? Should we allow the user to specify a range, such as in the print dialog of any Windows app?

And, if we did, none of these would be "default", would they?

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/Utils.java:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  public static List<Table> extractTablesFromPDF(PDDocument document, ExtractionAlgorithm algorithm) {
+    System.setProperty("java.awt.headless", "true");
```
Again, we probably shouldn't be doing this.

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/Utils.java:
```diff
@@ -0,0 +1,102 @@
+
+    ObjectExtractor objectExtractor = new ObjectExtractor(document);
+    PageIterator pages = objectExtractor.extract();
+    List<Table> tables= new ArrayList<>();
```
Nit: space before `=`.

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/Utils.java:
```diff
@@ -0,0 +1,102 @@
+
+  /**
+   * Returns a list of tables found in a given PDF document.  There are several extraction algorithms
+   * available and this function uses the default Basic Extraction Algorithm.
```
Maybe reference where this algorithm came from? What are the other algorithms?

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/Utils.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Utils {
```
As far as I can tell, the only utility here is the extractor. Should it be `TableExtractor` instead?

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/Utils.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    try {
+      objectExtractor.close();
```
How critical is this `close()`? Should it be in a `finally` block in case any of the above lines throws an exception?

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/Utils.java:
```diff
@@ -0,0 +1,102 @@
+    try {
+      objectExtractor.close();
+    } catch (Exception e) {
+      logger.debug("Error closing Object extractor.");
```
Should we be doing something useful with any exceptions, such as telling the user that the extract failed?

This line should say something about which file failed. Else, in a Drill that processed 100 PDF files, how will we know where to look for problems?

cgivre replied:
Changed to throw `UserException` if the table is unparsable.  This may actually be an interesting point.  One of the ways Pandas operates is that the user can choose the behavior when it can't parse data.   We may want to think about modifying Drill such that a user could decide "I want an error if Drill can't read the data" or "I want Drill to ignore bad data"

@dzamo Any thoughts here?

dzamo replied:
@cgivre I have just added a description of an option like that to the Drill 2.0 wiki page.

paul-rogers commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/Utils.java:
```diff
@@ -0,0 +1,102 @@
+    List<String> values = new ArrayList<>();
+
+    if (firstRow != null) {
+      for (int i =0; i < firstRow.size(); i++) {
```
Nit: space after `=`.

What happens if the rows have different sizes? Is that possible if we guessed wrong about the table? If the table merges cells? How should we handle such issues? Do we have tests for them?

cgivre replied:
We didn't test for all those use cases, but the underlying library (Tabula) actually has a pretty robust collection of tests that do cover a lot of what you're asking about.[1] 

In my manual testing, what I found was that if you had inconsistent tables, Tabula would basically produce some sort of mapping to columns and rows that Drill could read.   Not ideal, but neither is trying to extract data from PDFs.

[1]: https://github.com/tabulapdf/tabula-java/tree/master/src/test/java/technology/tabula

paul-rogers commented on contrib/format-pdf/src/main/resources/bootstrap-format-plugins.json:
```diff
@@ -0,0 +1,43 @@
+        }
+      }
+    },
+    "cp": {
```
Will we provide any PDFs built into Drill?

cgivre replied:
I didn't plan on it.  We can include one in the docs that people can d/l and try out.

paul-rogers commented on contrib/format-pdf/src/test/resources/logback-test.xml:
```diff
@@ -0,0 +1,57 @@
+    <level value="DEBUG" />
+    <appender-ref ref="STDOUT" />
+  </logger>
+</configuration>
```
Missing newline.

cgivre replied:
This was accidentally left in while I was writing this.  Will remove entirely.

paul-rogers commented on contrib/format-pdf/src/test/resources/logback-test.xml:
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Not clear if we want to do this. IIRC, Logback gets very confused if there are multiple config files on the class path: it prints a warning an arbitrarily picks one.

Why do we need a separate config file for this module?

cgivre replied:
I accidentally left this... it's now gone.

paul-rogers reviewed (changes requested):
Multiple comments: some architectural, some nits.

There are many example PDF files in the test directory. Not clear from their names what they are used for. Maybe add a "description.txt" or some such to explain how we use each for testing. That is:

* china.pdf: Verifies that Unicode (Chinese) characters are read correctly.
* ...

dzamo commented:
Copy that @paul-rogers, I [have incorporated your advice above into the Drill 2.0 wiki page](https://github.com/apache/drill/wiki/Drill-2.0-Proposal#project-structure-packaging-and-distribution).  I am hopeful that after 1.20 is out the door we can split out the new branches, and repos, for 2.0 (but it is of course the project elders that decide such things).  In my youthful exuberance I imagine it starting early next year!

I would think that getting the source code reorganised in a way that will carry us into the future should probably be the very first Drill 2.0 activity to take place.  Surely we would want our most experienced contributors to define such a task.

dzamo commented:
Dear PR author and reviewers.

This is a generic message to say that we would like to merge this PR in time for the 1.20 release.  Currently we're targeting a master branch freeze date of 2021-12-10 (10 Dec).  Please strive to complete development and review by this time, or indicate that the PR will need more time (and how much).

Thank you.

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/Utils.java:
```diff
@@ -0,0 +1,102 @@
+   * @return A list of the header rows
+   */
+  public static List<String> extractRowValues(Table table) {
+    List<RectangularTextContainer> firstRow = table.getRows().get(0);
```
Added null check.  I don't think it is possible, but the function now returns an empty list if table is null.

cgivre commented on contrib/format-pdf/src/test/java/org/apache/drill/exec/store/pdf/TestPdfFormat.java:
```diff
@@ -0,0 +1,229 @@
+      .addNullable("Apellido y Nombre", MinorType.VARCHAR)
+      .addNullable("Bloque político", MinorType.VARCHAR)
+      .addNullable("Provincia", MinorType.VARCHAR)
+      .addNullable("field_0", MinorType.VARCHAR)
```
Thx!

cgivre commented on contrib/format-pdf/src/test/java/org/apache/drill/exec/store/pdf/testPDFUtils.java:
```diff
@@ -0,0 +1,78 @@
+    assertEquals(values.size(), 11);
+  }
+
+
```
Fixed.

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+
+    // Prepare for reading
+    currentRowIndex = 1;  // Skip the first line if there are headers
+    currentTable = tables.get(startingTableIndex);
```
Added an exception with a good error message.  What would have happened is that the user would have received an `IndexOutOfBoundsException`.

cgivre commented on contrib/format-pdf/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## Data Model
+Since PDF files generally are not intended to be queried or read by machines, mapping the data to tables and rows is not a perfect process.  The PDF reader does support 
+provided schema. 
```
Good call!

cgivre commented on contrib/format-pdf/README.md:
```diff
@@ -0,0 +1,67 @@
+
+### Merging Pages
+The PDF reader reads tables from PDF files on each page.  If your PDF file has tables that span multiple pages, you can set the `combinePages` parameter to `true` and Drill 
+will merge all the tables in the PDF file.  You can also do this at query time with the `table()` function.
```
The way I envisioned this working was for a user to use the `table()` function and specify the table index at query time.  That way they can query as many incompatible tables as they want and not have to read ones they don't.

cgivre commented on contrib/format-pdf/README.md:
```diff
@@ -0,0 +1,67 @@
+will merge all the tables in the PDF file.  You can also do this at query time with the `table()` function.
+
+## Configuration
+To configure the PDF reader, simply add the information below to the `formats` section of a file base storage plugin. 
```
I meant to say file baseD storage plugin... but I added an example.

cgivre commented on contrib/format-pdf/README.md:
````diff
@@ -0,0 +1,67 @@
+}
+```
+The avaialable options are:
+* `extractHeaders`: Extracts the first row of any tables as the header row.  If set to false, Drill will assign column names of `field_0`, `field_1` to each column.
````
This is on a per-table basis.  Really it works the same way the CSV reader would work if you tried to query multiple files in a directory. 
In the example you cited, you'd get a table with columns, a | b | c | d | e.  The rows from table 1 would include `null` entries for the columns c | d | e and the rows from table 2 would contain nulls for columns a | b.

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+    // Get the tables
+    tables = Utils.extractTablesFromPDF(document);
+    populateMetadata();
+    logger.debug("Found {} tables", tables.size());
```
Removed...

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+
+  @Override
+  public void close() {
+    if (fsStream != null) {
```
Fixed.

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+
+    while(!rowWriter.isFull()) {
+      // Check to see if the limit has been reached
+      if (rowWriter.limitReached(maxRecords)) {
```
Good idea.  I'd prefer to do that in a separate PR.

paul-rogers replied:
I'd forgotten this comment. Limit was added to the EVF framework in the pending EVF V2 PR.

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+        .message("Failed to open open input file: %s", split.getPath().toString())
+        .addContext(e.getMessage())
+        .addContext(errorContext)
+        .build(logger);
```
Fixed

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+      // Check to see if the limit has been reached
+      if (rowWriter.limitReached(maxRecords)) {
+        return false;
+      } else if (currentRowIndex >= currentTable.getRows().size() &&
```
I cleaned this up by wrapping this in iterator classes.  This logic was here to support the page merge feature.  Basically... it works like this... a `Page` can contain 0 or more `Tables`.   If you take a look at the included file `schools.pdf` you'll see it is a 5 page spreadsheet.  The problem is that if you wanted to actually query this data, you'd want the whole thing, not an individual page.  So, what Drill can do is merge all the Pages (each containing 1 table) together. 

You would only want to use this feature in the case of multi-page spreadsheet PDFs.  The logic therefore is:
1. Read the current table until there are no more rows.
2. Get the next table (if there is one) 
3. Read it until there are no more rows
4. Go back to step 2.

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+    openFile();
+
+    // Get the tables
+    tables = Utils.extractTablesFromPDF(document);
```
Refactored to only open the desired table.  If the user has the `combineTables` set to true it does read all the tables in a given file at once.

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+    if (value == null) {
+      return;
+    }
+    writers.get(index).getWriter().setString(value);
```
Fixed (I think)

cgivre replied:
I refactored all this with a metadata class.

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+  private void addImplicitColumnsToSchema() {
+    metadataIndex = columns;
+    // Add to schema
+    addMetadataColumnToSchema("_page_count", MinorType.INT);
```
Done

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+    creationDate = info.getCreationDate();
+    modificationDate = info.getModificationDate();
+    trapped = info.getTrapped();
+    tableCount = tables.size();
```
Done

cgivre commented on contrib/format-pdf/src/main/java/org/apache/drill/exec/store/pdf/PdfBatchReader.java:
```diff
@@ -0,0 +1,549 @@
+      }
+      writers.get(i).load(row.get(i));
+    }
+    writeMetadata();
```
I was thinking about this... if you try to query an empty file of another type in Drill, would you get the implicit fields?  While I see your point, I'm not sure what the "correct" behavior should be.

cgivre commented on contrib/format-pdf/README.md:
```diff
@@ -0,0 +1,67 @@
+ * `_creation_date`
+ * `_modification_date`
+ * `_trapped`
+ * `_table_count`
```
> Here, I assume `_author` will repeat for every table row? That `_page_count` will increase and gives the page number? Or, is constant and will always be, say, 25 if that's the number of pages?

Page count is the total number of pages in the document.  All these metafields will be the same for every row.  Drill does this with the other file based readers and adds implicit fields, where if you query them, you'll get the same info for every row.  
> 
> Does `_table_count` stay fixed at the number of tables (5, say), or does it count tables (1, 2, 3, 4, 5)?

It was supposed to be the total number of tables in the document, however, I ended up removing it because in order to compute it, I had to load all the tables into memory. 

> 
> What is `_trapped`? Is there some PDF standard we could point to where this stuff is defined?

Yes.... I'll add that to the docs.  It has something to do with how a PDF doc is printed.

> 
> Actually, there is a larger question. If I'm exploring a pile of PDFs, I may want to extract metadata, such as the above. If I'm extracting tables, I don't need the metadata repeated per table row.
> 
> Should the plugin allow two query types? Either metadata or tables? Can this be done with some kind of Calcite trickery? `FROM "foo.pdf"` means the tables, `FROM "foo.pdf.metadata"` means the metadata?

It could but this is significantly more complex to do that.

> 
> Or, we check and if we see only metadata columns, we return one row per file rather than one row per table row?
> 
> Wouldn't something like this be useful to extract metadata from a PDF with no tables?

I'm debating this point... stand by.

cgivre replied:
I fixed this such that if a PDF has no tables, you can still get the metadata.

cgivre commented:
@paul-rogers 
Thanks for all your review.  I addressed all your comments (I think) and did the following:
* Added additional unit tests
* Refactored table list so that all tables are not read into memory if not requested
* Added iterator classes to avoid counters in the batch reader
* Moved metadata collection to separate class
* Refactored to allow a pdf with no tables to return metadata if requested
* Added config option for different extraction algorithms.
* General code cleanup

I removed all but one of the `System.env` calls and I'm a little stuck on this.  The reason I added this line is that when querying a PDF with Drill in embedded mode, it opens an additional java window.  This does not occur when running unit tests which makes for difficult debugging.   I'm going to keep digging into this, but I was wondering if you could take a look at the rest of the revisions in the mean time?   The issue seems to be in either Tabula or PdfBox, which are the underlying libraries that read the PDF file. Thanks!

pjfanning commented:
@cgivre new pdfbox release: https://downloads.apache.org/pdfbox/2.0.25/RELEASE-NOTES.txt

cgivre commented:
> @cgivre new pdfbox release: https://downloads.apache.org/pdfbox/2.0.25/RELEASE-NOTES.txt

Thanks @pjfanning !  Unfortunately, that didn't solve the issue with opening up the java window.  I emailed the PDFbox user list and hopefully they can give me some suggestions.

cgivre commented:
@paul-rogers 
This PR is now ready for review.  I did some research into the `headless` option and why specifically I was not getting this issue when I ran unit tests, and only when I ran manual queries.  So, the issue was caused when PDFBox opens a document, it calls some AWT libraries that it needs for deciphering PDF files.  (Also probably for writing PDF files)   In any event, this is what was causing the window to open. 

When Drill runs the unit tests, we actually have the `headless` option specified as a Java option.  So, I moved this setting to `drill-config.sh` and included a comment about why it is there.  If someone wants to remove it for some reason, they can and they can disable the PDF functionality simply by not including it in the configs for the storage plugin.  This seems like the right place for this to live. 

I believe that I've addressed all your other comments as well.  Thank you as always for the thorough review!  It's ready for another look.

paul-rogers reviewed (approved):
Still not convinced that this should be part of the core Drill code: reading PDF files pushes Drill in the direction of a small-scale, query-anything tool rather than a large-scale distributed query tool. Folks won't, at the same time, run 100 Drill nodes and muck with PDF files: PDF is a horrible format to use for data at a scale.

Further, I'm not convinced the SQL and JSON are the best format for working out how to parse a PDF file. If the user is naive enough to use PDF as their input format, they're not going to be savvy enough to figure out the ad-hoc, clunky way they have to configure stuff using SQL and JSON.

That said, that conversation was already had, and we're proceeding anyway. Given that, the code look good. So, +1.
